### PR TITLE
fix some scipy segfaults

### LINKED ISSUE #1275: Wrong refcount increasing?
Hi @kmod 

Remember the issues that because an additional refcount, NumPy throw the exception in below in https://github.com/Daetalus/numpy/blob/master/numpy/core/src/multiarray/shape.c#L92:
```
ValueError: cannot resize an array that references or is referenced
by another array in this way.  Use the resize function
```

I only can track down the place where got the additional refcount. It is here: 
https://github.com/dropbox/pyston/blob/master/src/runtime/objmodel.cpp#L4288
After that, the array(`0x7fffde4da640`)'s refcount is 3.

Sorry, the information is limited(no snippet which can re-produce the error yet) for now. Maybe I should file this issue after other scipy problems get solved. 

Here is the backtrace:
```
#0  PyArray_Resize (self=0x7fffde4da640, newshape=0x7ffffffaf350, refcheck=1, order=NPY_CORDER) at numpy/core/src/multiarray/shape.c:41
#1  0x00007ffff34dde0c in array_resize (self=0x7fffde4da640, args=0x7ffff485c220, kwds=0x0) at numpy/core/src/multiarray/methods.c:1062
#2  0x000000000089e084 in pyston::Box* pyston::methodDescrTppCall<(pyston::ExceptionStyle)1>(pyston::Box*, pyston::CallRewriteArgs*, pyston::ArgPassSpec, pyston::Box*, pyston::Box*, pyston::Box*, pyston::Box**, std::vector<pyston::BoxedString*, std::allocator<pyston::BoxedString*> > const*)::{lambda(pyston::CallRewriteArgs*, pyston::Box*, pyston::Box*, pyston::Box*, pyston::Box**)#1}::operator()(pyston::CallRewriteArgs*, pyston::Box*, pyston::Box*, pyston::Box*, pyston::Box**) const (this=0x7ffffffb01a0, rewrite_args=0x0, arg1=0x7fffde4da640, arg2=0x7fffde4d90d0, arg3=0x0, args=0x0) at ../../src/runtime/descr.cpp:334
#3  0x000000000089dcf0 in llvm::function_ref<pyston::Box* (pyston::CallRewriteArgs*, pyston::Box*, pyston::Box*, pyston::Box*, pyston::Box**)>::callback_fn<pyston::Box* pyston::methodDescrTppCall<(pyston::ExceptionStyle)1>(pyston::Box*, pyston::CallRewriteArgs*, pyston::ArgPassSpec, pyston::Box*, pyston::Box*, pyston::Box*, pyston::Box**, std::vector<pyston::BoxedString*, std::allocator<pyston::BoxedString*> > const*)::{lambda(pyston::CallRewriteArgs*, pyston::Box*, pyston::Box*, pyston::Box*, pyston::Box**)#1}>(long, pyston::CallRewriteArgs*, pyston::Box*, pyston::Box*, pyston::Box*, pyston::Box**) (
    callable=140737488028064, params=0x0, params=0x0, params=0x0, params=0x0, params=0x0) at /home/sun/pyston_deps/llvm-trunk/include/llvm/ADT/STLExtras.h:74
#4  0x00000000009076fc in llvm::function_ref<pyston::Box* (pyston::CallRewriteArgs*, pyston::Box*, pyston::Box*, pyston::Box*, pyston::Box**)>::operator()(pyston::CallRewriteArgs*, pyston::Box*, pyston::Box*, pyston::Box*, pyston::Box**) const (this=0x7ffffffafff8, params=0x0, params=0x0, params=0x0, params=0x0, params=0x0) at /home/sun/pyston_deps/llvm-trunk/include/llvm/ADT/STLExtras.h:87
#5  0x0000000000907591 in pyston::rearrangeArgumentsAndCallInternal(pyston::ParamReceiveSpec, pyston::ParamNames const*, pyston::FuncNameGetter, pyston::Box**, pyston::CallRewriteArgs*, pyston::ArgPassSpec, pyston::Box*, pyston::Box*, pyston::Box*, pyston::Box**, std::vector<pyston::BoxedString*, std::allocator<pyston::BoxedString*> > const*, llvm::function_ref<pyston::Box* (pyston::CallRewriteArgs*, pyston::Box*, pyston::Box*, pyston::Box*, pyston::Box**)>) (paramspec=..., param_names=0x0, func_name_cb=..., defaults=0x0, rewrite_args=0x0, argspec=..., arg1=0x7fffde5eed08, arg2=0x0, arg3=0x0, args=0x0, keyword_names=0x0, 
    continuation=...) at ../../src/runtime/objmodel.cpp:4592
#6  0x0000000000908552 in pyston::rearrangeArgumentsAndCall(pyston::ParamReceiveSpec, pyston::ParamNames const*, char const*, pyston::Box**, pyston::CallRewriteArgs*, pyston::ArgPassSpec, pyston::Box*, pyston::Box*, pyston::Box*, pyston::Box**, std::vector<pyston::BoxedString*, std::allocator<pyston::BoxedString*> > const*, llvm::function_ref<pyston::Box* (pyston::CallRewriteArgs*, pyston::Box*, pyston::Box*, pyston::Box*, pyston::Box**)>) (paramspec=..., param_names=0x0, func_name=0x7ffff35a0f3c "resize", defaults=0x0, rewrite_args=0x0, argspec=..., arg1=0x7fffde5eed08, arg2=0x0, arg3=0x0, args=0x0, keyword_names=0x0, 
    continuation=...) at ../../src/runtime/objmodel.cpp:4600
#7  0x000000000089dc18 in pyston::methodDescrTppCall<(pyston::ExceptionStyle)1> (_self=0x7ffff7e15e70, rewrite_args=0x0, argspec=..., arg1=0x7fffde5eed08, arg2=0x0, arg3=0x0, args=0x0, keyword_names=0x0)
    at ../../src/runtime/descr.cpp:386
#8  0x000000000089d635 in pyston::methodDescrTppCall<(pyston::ExceptionStyle)0> (_self=0x7ffff7e15e70, rewrite_args=0x0, argspec=..., arg1=0x7fffde5eed08, arg2=0x0, arg3=0x0, args=0x0, keyword_names=0x0)
    at ../../src/runtime/descr.cpp:234
#9  0x00000000006b67a1 in pyston::ExceptionSwitchableFunction<pyston::Box*, pyston::Box*, pyston::CallRewriteArgs*, pyston::ArgPassSpec, pyston::Box*, pyston::Box*, pyston::Box*, pyston::Box**, std::vector<pyston::BoxedString*, std::allocator<pyston::BoxedString*> > const*>::call<(pyston::ExceptionStyle)0> (this=0x1fc3070 <PyMethodDescr_Type+496>, args=0x0, args=0x0, args=0x0, args=0x0, args=0x0, args=0x0, args=0x0, 
    args=0x0) at ../../src/core/types.h:119
// SciPy back traces
```

edit: I took a look at https://github.com/dropbox/pyston/pull/1176/commits/6e7da64dc889092566e307a01e6578ad614748c7, but didn't find anything...

### PR DESCRIPTION
**Pull request by Daetalus**
closes #1275

### CODE REVIEW

**Comment by Daetalus** [comment]
@undingen Would you mind to take a look if you have time, please? Vielen Dank!

File: src/runtime/objmodel.cpp [hunk 1]
```diff
@@ -6479,7 +6479,9 @@ extern "C" void setitem(Box* target, Box* slice, Box* value) {
             RewriterVar* r_cls = r_obj->getAttr(offsetof(Box, cls));
             RewriterVar* r_m = r_cls->getAttr(offsetof(BoxedClass, tp_as_mapping));
             r_m->addGuardNotEq(0);
-            rewriter->call(true, (void*)setitemHelper, r_obj, r_slice, r_value);
+            r_m->addAttrGuard(offsetof(PyMappingMethods, mp_ass_subscript), (intptr_t)m->mp_ass_subscript);
+            RewriterVar* r_ret = rewriter->call(true, (void*)m->mp_ass_subscript, r_obj, r_slice, r_value);
+            rewriter->checkAndThrowCAPIException(r_ret, -1, assembler::MovType::L);
             rewriter->commit();
         }
```
Comment (undingen) [inline src/runtime/objmodel.cpp]: could you please also remove the `setitemHelper` function and replace the call with
```
int ret = m->mp_ass_subscript(target, slice, value);
if (ret == -1)
   throwCAPIException();
```

File: src/asm_writing/rewriter.h [hunk 2]
```diff
@@ -578,9 +578,12 @@ class Rewriter : public ICSlotRewrite::CommitHook {
 
     void assertConsistent() {
```
Comment (undingen) [inline src/asm_writing/rewriter.h]: I think it's better to just add a `if (failed) return;` at the beginning of the function

File: src/asm_writing/rewriter.h [hunk 3]
```diff
@@ -577,6 +577,9 @@ class Rewriter : public ICSlotRewrite::CommitHook {
     void _callOptimalEncoding(assembler::Register tmp_reg, void* func_addr);
 
     void assertConsistent() {
+
+        if (!failed)
```
Comment (undingen) [inline src/asm_writing/rewriter.h]: you swapped the conditional here

File: src/runtime/objmodel.cpp [hunk 4]
```diff
@@ -4285,9 +4269,28 @@ Box* rearrangeArgumentsAndCallInternal(ParamReceiveSpec paramspec, const ParamNa
     int varargs_to_positional = std::min((int)varargs_size, paramspec.num_args - positional_to_positional);
     for (int i = 0; i < varargs_to_positional; i++) {
         assert(!rewrite_args && "would need to be handled here");
-        getArg(i + positional_to_positional, oarg1, oarg2, oarg3, oargs) = incref(PySequence_Fast_GET_ITEM(varargs, i));
+        getArg(i + positional_to_positional, oarg1, oarg2, oarg3, oargs) = PySequence_Fast_GET_ITEM(varargs, i);
     }
 
+    auto clear_refs = [&]() {
```
Comment (undingen) [inline src/runtime/objmodel.cpp]: very minor but could you please move the `// Clear any increfs we did for when we throw an exception:` comment to here (in front of `auto clear_refs = [&]() {`)

**Comment by undingen** [comment]
very minor comment, after that it should be ready for merging

**Comment by undingen** [comment]
thanks for the patch! :+1:

**Pull request merged**